### PR TITLE
Issue #2532 Better handling of tokens in parser

### PR DESCRIPTION
**Pull request by gregw**
Issue #2532 Better handling of tokens in parser
Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java [hunk 1]
```diff
@@ -1881,11 +1961,12 @@ public default void onComplianceViolation(HttpCompliance compliance, HttpComplia
     @SuppressWarnings("serial")
     private static class IllegalCharacterException extends BadMessageException
     {
-        private IllegalCharacterException(State state,byte ch,ByteBuffer buffer)
+        private IllegalCharacterException(State state,HttpTokens.Token token,ByteBuffer buffer)
         {
-            super(400,String.format("Illegal character 0x%X",ch));
             // Bug #460642 - don't reveal buffers to end user
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: I think this line should be removed, it references a bugzilla ID that we don't use anymore, and there is `git blame` to figure out these things.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java [hunk 2]
```diff
@@ -569,19 +520,28 @@ else if (_responseHandler!=null)
...
+                case TCHAR:
+                case VCHAR:
+                {
+                    _string.setLength(0);
+                    _string.append(t.getChar());
+                    setState(_requestHandler!=null?State.METHOD:State.RESPONSE_VERSION);
+                    return false;
+                }
+                case OTEXT:
+                    throw new BadMessageException();
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java]: Add error message to `BadMessageException`.

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpTokens.java [hunk 3]
```diff
@@ -34,5 +35,167 @@
...
+                    else if (b>=0x80) // OBS
+                        TOKENS[b] = new Token((byte)b,Type.OTEXT);
+                    else
+                        TOKENS[b] = new Token((byte)b,Type.CNTL);
+            }
+        }
+    }
+    
+    
+    public static void main(String... args)
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpTokens.java]: Remove this `main` method.

File: jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartParser.java [hunk 4]
```diff
@@ -694,16 +723,17 @@ default void earlyEOF()
         {
         }
     }
-    
+
     /* ------------------------------------------------------------------------------- */
     @SuppressWarnings("serial")
-    private static class IllegalCharacterException extends IllegalArgumentException
+    private static class IllegalCharacterException extends BadMessageException
```
Comment (sbordet) [inline jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartParser.java]: This class is used in both `HttpParser` and here, no point in duplicating, make it public in its own file.

Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartParser.java]: The class is similar, but subtly different.  It has a different parent class and a different state that it reports

**Comment by lachlan-roberts** [comment]
Here is a version of the BadHeaderNames test for MultiPartParserTest
https://gist.github.com/lachlan-roberts/8b8279a82e94636694bb04c17c7b745b

**Pull request merged**